builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,219 @@
+				Default:     30,
+				Description: `The length of time used to generate a counter for the TOTP token calculation.`,
+			},
+
+			"algorithm": {
+				Type:        framework.TypeString,
+				Default:     "SHA1",
+				Description: `The hashing algorithm used to generate the TOTP token.`,
+			},
+
+			"digits": {
+				Type:        framework.TypeInt,
+				Default:     6,
+				Description: `The number of digits in the generated TOTP token.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("role/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	// Return values of role
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"issuer":       role.Issuer,
+			"account_name": role.Account_Name,
+			"period":       role.Period,
+			"algorithm":    role.Algorithm,
+			"digits":       role.Digits,
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	key := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	account_name := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+
+	// Enforce input value requirements
+	if key == "" {
+		return logical.ErrorResponse("The key value is required."), nil
+	}
+
+	_, err := base32.StdEncoding.DecodeString(key)
+
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf(
+			"Invalid key value: %s", err)), nil
+	}
+
+	if period < 0 {
+		return logical.ErrorResponse("The period value must be greater than zero."), nil
+	}
+
+	// Set optional parameters if neccessary
+	if period == 0 {
+		period = 30
+	}
```

Comment:
This can be removed now that there is a default and if the above branch is changed to `<=`